autopush-common/src/db/bigtable/bigtable_client/mod.rs

```diff
@@ -509,15 +509,22 @@ impl BigTableClientImpl {
         let resp = retry_policy(self.settings.retry_count)
             .retry_if(
                 || async {
-                    bigtable
+                    let resp: grpcio::ClientSStreamReceiver<bigtable::ReadRowsResponse> = bigtable
                         .conn
                         .read_rows_opt(&req, call_opts(self.metadata.clone()))
+                        .inspect_err(|e| {
+                            warn!("🉑 Read Rows failed: {:?}", &e);
+                        })?;
+                    merge::RowMerger::process_chunks(resp).await.map_err(|e| {
+                        warn!("🉑 Process Chunks failed {:?}", e);
+                        grpcio::Error::RemoteStopped
```

Comment:
Or maybe it's as easy as special casing `InvalidRowResponse` here, pulling out the inner `grpcio::Error` and propagating that instead?

Reply:
If that does work, we might need to begin capturing inner `Backtrace`s in `BigtableError` because it can already be a bit confusing determining what inner call triggered an error -- this change would probably make that worse.

Reply:
Looking at this a bit more. `process_chunks` returns a `error::BigTableError` some of which can be `grpcio::Error` types. For those, we should return the embedded GRPCIO error to allow the `retry_if` loop to complete. 
For other BigTable related errors, however, we need to convert them to some `grpcio::Error` type.
The only outlier error that `process_chunks` could generate would be `BigTableError::InvalidChunk`, which returns a string message. We could translate that as a `grpcio::Error::RpcFailure(RpcStatus::with_message(RpcStatusCode::UNKNOWN, e.to_string()))` which, while not great, would at least capture the message should we get those, since `retry_if` is looking for `grpcio::Error` messages. 
I'm not sure about capturing the backtrace, since we're transmuting the error type here.

Reply:
I could be missing something but I only see `process_chunks`/`merge` returning that single `BigTableError::InvalidRowResponse` and every other error returned is `InvalidChunk`.

So matching on `error::BigTableError::GRPC(e) | error::BigTableError::Read(e)` wouldn't do anything?

I guess using `RpcFailure` for `InvalidChunk` is fine. I suppose if we really cared we could also make another version of `retryable_error` that works on `BigTableError` for this case.

I think you're right we don't need the backtrace, just as long as it's obvious what was the root cause of the error was. We only use `read_rows` in a couple places and it's the sole db call in those methods, so it should hopefully still be obvious what triggered a certain `grpcio::Error`